src/java/htsjdk/samtools/util/AsyncBufferedIterator.java

```diff
@@ -0,0 +1,261 @@
+     * @param bufferSize size of each read-ahead buffer. A larger size will increase both throughput and latency.
+     * @param bufferCount number of read-ahead buffers
+     * @param threadName background thread name. A name will be automatically generated if this parameter is null.
+     */
+    public AsyncBufferedIterator(final Iterator<T> iterator, final int bufferSize, final int bufferCount, final String threadName) {
+        if (iterator == null) throw new IllegalArgumentException("iterator cannot be null");
+        if (bufferCount <= 0) throw new IllegalArgumentException("Must use at least 1 buffer.");
+        if (bufferSize <= 0) throw new IllegalArgumentException("Buffer size must be at least 1 record.");
+        this.underlyingIterator = iterator;
+        this.buffer = new ArrayBlockingQueue<>(bufferCount);
+        this.bufferSize = bufferSize;
+        int threadNumber = threadsCreated.incrementAndGet();
+        this.backgroundThread = new Thread(new Runnable() {
+                @Override
+                public void run() {
+                    backgroundRun();
+                }
+            }, threadName != null ? threadName : getThreadNamePrefix() + threadNumber);
+        this.backgroundThread.setDaemon(true);
+        log.debug("Starting thread " + this.backgroundThread.getName());
+        this.backgroundThread.start();
+    }
+
+    protected String getThreadNamePrefix() {
+        return AsyncBufferedIterator.class.getSimpleName();
+    }
+
+    @Override
+    public void close() {
+        if (backgroundThread != null) {
+            try {
+                backgroundThread.interrupt();
+                buffer.clear();
+                backgroundThread.join();
+            } catch (InterruptedException ie) {
+                throw new RuntimeException("Interrupted waiting for background thread to complete", ie);
+            } finally {
+                CloserUtil.close(underlyingIterator);
+                backgroundThread = null;
+                currentBlock = null;
+            }
+        }
+    }
+
+    @Override
+    public boolean hasNext() {
+        if (backgroundThread == null) {
+            throw new IllegalStateException("iterator has been closed");
+        }
+        if (currentBlock.hasNext()) return true;
+        
+        // rethrow any exceptions raised on the background thread
+        // at the point the exception would have been encountered
+        // if we had performed synchronous iteration
+        Throwable t = currentBlock.getException();
+        if (t != null) {
+            if (t instanceof Error) {
+                throw (Error) t;
+            } else if (t instanceof RuntimeException) {
+                throw (RuntimeException) t;
+            } else {
+                throw new RuntimeException(t);
+            }
+        }
+        if (currentBlock.isEndOfStream()) return false;
+        try {
+            // try loading next block
+            currentBlock = buffer.take();
```

Comment:
what happens if the main thread is blocked here on an empty list and the worker thread blows up somewhere in `readAhead` or `backgroundRun` and dies? Will the main thread block forever?